api/v1beta2/imagerepository_types.go

```diff
@@ -0,0 +1,200 @@
+	"github.com/fluxcd/pkg/apis/meta"
+)
+
+const ImageRepositoryKind = "ImageRepository"
+const ImageRepositoryFinalizer = "finalizers.fluxcd.io"
+
+// ImageRepositorySpec defines the parameters for scanning an image
+// repository, e.g., `fluxcd/flux`.
+type ImageRepositorySpec struct {
+	// Image is the name of the image repository
+	// +required
+	Image string `json:"image,omitempty"`
+	// Interval is the length of time to wait between
+	// scans of the image repository.
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ms|s|m|h))+$"
+	// +required
+	Interval metav1.Duration `json:"interval,omitempty"`
+
+	// Timeout for image scanning.
+	// Defaults to 'Interval' duration.
+	// +kubebuilder:validation:Type=string
+	// +kubebuilder:validation:Pattern="^([0-9]+(\\.[0-9]+)?(ms|s|m))+$"
+	// +optional
+	Timeout *metav1.Duration `json:"timeout,omitempty"`
+
+	// SecretRef can be given the name of a secret containing
+	// credentials to use for the image registry. The secret should be
+	// created with `kubectl create secret docker-registry`, or the
+	// equivalent.
+	// +optional
+	SecretRef *meta.LocalObjectReference `json:"secretRef,omitempty"`
+
+	// ServiceAccountName is the name of the Kubernetes ServiceAccount used to authenticate
+	// the image pull if the service account has attached pull secrets.
+	// +kubebuilder:validation:MaxLength=253
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty"`
+
+	// CertSecretRef can be given the name of a secret containing
+	// either or both of
+	//
+	//  - a PEM-encoded client certificate (`certFile`) and private
+	//  key (`keyFile`);
+	//  - a PEM-encoded CA certificate (`caFile`)
+	//
+	//  and whichever are supplied, will be used for connecting to the
+	//  registry. The client cert and key are useful if you are
+	//  authenticating with a certificate; the CA cert is useful if
+	//  you are using a self-signed server certificate.
+	// +optional
+	CertSecretRef *meta.LocalObjectReference `json:"certSecretRef,omitempty"`
+
+	// This flag tells the controller to suspend subsequent image scans.
+	// It does not apply to already started scans. Defaults to false.
+	// +optional
+	Suspend bool `json:"suspend,omitempty"`
+
+	// AccessFrom defines an ACL for allowing cross-namespace references
+	// to the ImageRepository object based on the caller's namespace labels.
+	// +optional
+	AccessFrom *acl.AccessFrom `json:"accessFrom,omitempty"`
+
+	// ExclusionList is a list of regex strings used to exclude certain tags
+	// from being stored in the database.
+	// +kubebuilder:default:={"^.*\\.sig$"}
+	// +kubebuilder:validation:MaxItems:=25
+	// +optional
+	ExclusionList []string `json:"exclusionList,omitempty"`
+
+	// The provider used for authentication, can be 'aws', 'azure', 'gcp' or 'generic'.
+	// When not specified, defaults to 'generic'.
+	// +kubebuilder:validation:Enum=generic;aws;azure;gcp
+	// +kubebuilder:default:=generic
```

Comment:
The defaults only work at admission, so none of the existing objects will get them after conversion, only newly created objects get the defaults. IMO we shouldn't be using defaults if we care about backwards compatibility and we want users to be able to rollback. I very much regret having a default for `gitImplementation` which is now deprecated and no longer use, we are now in a situation where we can't delete it from the CRD because everyone has a value for it in etcd.

Reply:
> If you check my last comment above about upgrade testing, the defaults are added to existing objects.

I don't see that on my test cluster (v1.25.3).

Before upgrade:

```yaml
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  creationTimestamp: "2023-02-08T15:45:54Z"
  generation: 1
  name: podinfo
  namespace: default
  resourceVersion: "591"
  uid: 200bca96-f5d0-4ec3-bd75-0900d385c59b
spec:
  accessFrom:
    namespaceSelectors:
    - matchLabels:
        kubernetes.io/metadata.name: flux-system
  image: ghcr.io/stefanprodan/podinfo
  interval: 1m0s
status:
  observedGeneration: -1
```

After upgrade there no defaults:

```yaml
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImageRepository
metadata:
  creationTimestamp: "2023-02-08T15:45:54Z"
  generation: 1
  name: podinfo
  namespace: default
  resourceVersion: "591"
  uid: 200bca96-f5d0-4ec3-bd75-0900d385c59b
spec:
  accessFrom:
    namespaceSelectors:
    - matchLabels:
        kubernetes.io/metadata.name: flux-system
  image: ghcr.io/stefanprodan/podinfo
  interval: 1m0s
status:
  observedGeneration: -1
```

But if I apply the object again, then the defaults are set because it happens at admission:

```yaml
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImageRepository
metadata:
  creationTimestamp: "2023-02-08T15:45:54Z"
  generation: 1
  name: podinfo
  namespace: default
  resourceVersion: "1086"
  uid: 200bca96-f5d0-4ec3-bd75-0900d385c59b
spec:
  accessFrom:
    namespaceSelectors:
    - matchLabels:
        kubernetes.io/metadata.name: flux-system
  exclusionList:
  - ^.*\.sig$
  image: ghcr.io/stefanprodan/podinfo
  interval: 1m0s
  provider: generic
status:
  observedGeneration: -1
```

Reply:
I can't reproduce the dry-run error with Kubernetes Kind, what people reported for NC was specific to EKS. So we can leave the defaults in and deal with EKS bugs later.